lightning/src/ln/chan_utils.rs

```diff
@@ -1236,8 +1266,78 @@ fn script_for_p2wpkh(key: &PublicKey) -> Script {
+	use bitcoin::secp256k1::{PublicKey, SecretKey, Secp256k1};
+	use util::test_utils;
+	use chain::keysinterface::{KeysInterface, BaseSign};
+	use bitcoin::Network;
+
+	#[test]
+	fn test_anchors() {
+		let secp_ctx = Secp256k1::new();
+
+		let seed = [42; 32];
+		let network = Network::Testnet;
+		let keys_provider = test_utils::TestKeysInterface::new(&seed, network);
+		let signer = keys_provider.get_channel_signer(false, 3000);
+		let counterparty_signer = keys_provider.get_channel_signer(false, 3000);
+		let delayed_payment_base = &signer.pubkeys().delayed_payment_basepoint;
+		let per_commitment_secret = SecretKey::from_slice(&hex::decode("1f1e1d1c1b1a191817161514131211100f0e0d0c0b0a09080706050403020100").unwrap()[..]).unwrap();
+		let per_commitment_point = PublicKey::from_secret_key(&secp_ctx, &per_commitment_secret);
+		let htlc_basepoint = &signer.pubkeys().htlc_basepoint;
+		let holder_pubkeys = signer.pubkeys();
+		let counterparty_pubkeys = counterparty_signer.pubkeys();
+		let keys = TxCreationKeys::derive_new(&secp_ctx, &per_commitment_point, delayed_payment_base, htlc_basepoint, &counterparty_pubkeys.revocation_basepoint, &counterparty_pubkeys.htlc_basepoint).unwrap();
+		let channel_parameters = ChannelTransactionParameters {
+			holder_pubkeys: holder_pubkeys.clone(),
+			holder_selected_contest_delay: 0,
+			is_outbound_from_holder: false,
+			counterparty_parameters: Some(CounterpartyChannelTransactionParameters { pubkeys: counterparty_pubkeys.clone(), selected_contest_delay: 0 }),
+			funding_outpoint: Some(chain::transaction::OutPoint { txid: Default::default(), index: 0 })
+		};
+
+		let mut htlcs_with_aux: Vec<(_, ())> = Vec::new();
+
+		let tx = CommitmentTransaction::new_with_auxiliary_htlc_data(
+			0, 1000, 2000,
+			false,
+			holder_pubkeys.funding_pubkey,
+			counterparty_pubkeys.funding_pubkey,
+			keys.clone(), 1,
+			&mut htlcs_with_aux, &channel_parameters.as_holder_broadcastable()
+		);
+		assert_eq!(tx.built.transaction.output.len(), 2);
+
+		let tx = CommitmentTransaction::new_with_auxiliary_htlc_data(
+			0, 1000, 2000,
+			true,
+			holder_pubkeys.funding_pubkey,
+			counterparty_pubkeys.funding_pubkey,
+			keys.clone(), 1,
+			&mut htlcs_with_aux, &channel_parameters.as_holder_broadcastable()
+		);
+		assert_eq!(tx.built.transaction.output.len(), 4);
+		let tx = CommitmentTransaction::new_with_auxiliary_htlc_data(
+			0, 3000, 0,
+			true,
+			holder_pubkeys.funding_pubkey,
+			counterparty_pubkeys.funding_pubkey,
+			keys.clone(), 1,
+			&mut htlcs_with_aux, &channel_parameters.as_holder_broadcastable()
+		);
+		assert_eq!(tx.built.transaction.output.len(), 2);
+		let tx = CommitmentTransaction::new_with_auxiliary_htlc_data(
+			0, 0, 3000,
+			true,
+			holder_pubkeys.funding_pubkey,
+			counterparty_pubkeys.funding_pubkey,
+			keys.clone(), 1,
+			&mut htlcs_with_aux, &channel_parameters.as_holder_broadcastable()
+		);
+		assert_eq!(tx.built.transaction.output.len(), 2);
```

Comment:
I think there is a missing test case when we have `!htlcs_with_aux.is_empty()` ?

```
diff --git a/lightning/src/ln/chan_utils.rs b/lightning/src/ln/chan_utils.rs
index 20734f95..e226dc84 100644
--- a/lightning/src/ln/chan_utils.rs
+++ b/lightning/src/ln/chan_utils.rs
@@ -1268,7 +1268,8 @@ mod tests {
        use super::CounterpartyCommitmentSecrets;
        use ::{hex, chain};
        use prelude::*;
-       use ln::chan_utils::{CommitmentTransaction, TxCreationKeys, ChannelTransactionParameters, CounterpartyChannelTransactionParameters};
+       use ln::PaymentHash;
+       use ln::chan_utils::{CommitmentTransaction, TxCreationKeys, ChannelTransactionParameters, CounterpartyChannelTransactionParameters, HTLCOutputInCommitment};
        use bitcoin::secp256k1::{PublicKey, SecretKey, Secp256k1};
        use util::test_utils;
        use chain::keysinterface::{KeysInterface, BaseSign};
@@ -1298,6 +1299,14 @@ mod tests {
                        funding_outpoint: Some(chain::transaction::OutPoint { txid: Default::default(), index: 0 })
                };
 
+               let payment_hash = PaymentHash([42; 32]);
+               let htlc_info = HTLCOutputInCommitment {
+                       offered: false,
+                       amount_msat: 1,
+                       cltv_expiry: 100,
+                       payment_hash,
+                       transaction_output_index: Some(3),
+               };
                let mut htlcs_with_aux: Vec<(_, ())> = Vec::new();
 
                let tx = CommitmentTransaction::new_with_auxiliary_htlc_data(
@@ -1337,6 +1346,15 @@ mod tests {
                        &mut htlcs_with_aux, &channel_parameters.as_holder_broadcastable()
                );
                assert_eq!(tx.built.transaction.output.len(), 2);
+               let tx = CommitmentTransaction::new_with_auxiliary_htlc_data(
+                       0, 3000, 0,
+                       true,
+                       holder_pubkeys.funding_pubkey,
+                       counterparty_pubkeys.funding_pubkey,
+                       keys.clone(), 1,
+                       &mut vec![(htlc_info, ())], &channel_parameters.as_holder_broadcastable()
+               );
+               assert_eq!(tx.built.transaction.output.len(), 4);
        }
 
        #[test]
```

Reply:
Good point.  Slightly modified the test to generate a non-dust HTLC output (in case we start sanity checking that in the future) and pass `None` for the HTLC tx index (it should be populated by the commitment tx constructor, but we don't check).